### ISSUE TITLE
customize Server request 

### ISSUE THREAD

**Issue by Bofuser** [carlrobertoh/CodeGPT#90]
When setting API keys, could you please add a customize Server function like this plug-in, allowing me to customize and call API keys of my own server?
![image](https://user-images.githubusercontent.com/51036600/233892281-f6dde3b8-dcc4-4a4c-be4f-e80ecdd21475.png)

 I would be very grateful if you can add this function.

**Comment by carlrobertoh** [comment]
Hi, yes it is possible. 
Shouldn't be a huge change, will try to add it some day during this week.

**Comment by renjfk** [comment]
@karolisKTri Would it work out of the box by just updating endpoint? I have created https://github.com/carlrobertoh/CodeGPT/issues/93 without checking in detail but if that's the case then it's great!

**Comment by renjfk** [comment]
I see, bummers. Well I'm planning to contribute Azure OpenAI services support but for now waiting for internal reaction from the company that I'm working for so I can allocate some work time for this contribution.

**Comment by carlrobertoh** [comment]
Changes are required in both openai-client lib and the plugin itself. 
@renjfk Could you hold off with the implementation until I add the Azure client support to the library?

Furthermore, I don't have the resources to test the logic, so it would be highly appreciated if anyone is willing to give me access to the Azure service solely for testing purposes.

**Comment by renjfk** [comment]
@carlrobertoh I can do that as well, contributing support to library that is.

**Comment by shalak** [comment]
@carlrobertoh I can do some testing for you, let me know once you have the branch with customized build.

**Comment by carlrobertoh** [comment]
@renjfk I pushed the lib changes, I didn't if you were already working on it or was just planning to, sorry about that.

@shalak I'd highly appreciate if you could test the new version of the openai-client lib:
```xml
<dependency>
    <groupId>ee.carlrobert</groupId>
    <artifactId>openai-client</artifactId>
    <version>1.0.10</version>
</dependency>
```

Example:
```java
new OpenAIClient.Builder("API_KEY")
    .buildAzureChatCompletionClient(
        new AzureClientRequestParams("resource-name", "deployment-id", "api-version"))
    .stream(
        new ChatCompletionRequest.Builder(List.of(new ChatCompletionMessage("user", "Hello there!")))
            .setModel(ChatCompletionModel.GPT_3_5)
            .build(),
        new CompletionEventListener() {
          @Override
          public void onMessage(String message) {
            System.out.println(message);
          }
        });
```

**Comment by shalak** [comment]
@carlrobertoh The `ChatCompletionRequest` works perfectly! :) 

I understand that the [DashboardClient](https://github.com/carlrobertoh/openai-client/commit/997d148a234877f87e785a6f293de4c60f12830d#diff-aff2c5337d981c6a42fc73ea6a6b3903c5a234d0e4941f99685e76f13ee3012aR14) is not applicable to Azure deployments, right?

**Comment by carlrobertoh** [comment]
Cool! Thank you!

I'm not sure, I didn't find any documentation for it.

**Comment by shalak** [comment]
@carlrobertoh I noticed one thing - [over here](https://github.com/carlrobertoh/openai-client/blob/master/src/main/java/ee/carlrobert/openai/client/completion/CompletionEventSourceListener.java#L80) with Azure Services when the API key is invalid, we don't get any `responseError` - it's null. But the `response` itself has code 401 and `body.string()` looks something like this:
```json
{
    "message": "Token is invalid...",
    "statusCode": 401
}
```

So there's room to improvement in that area. The biggest issue here is that user is not notified about error, nothing shows up in the chat window.

**Comment by carlrobertoh** [comment]
@shalak Thank you for contributing! I will try to review the PR as soon as possible.

FYI: I fixed the API error deserialization, you can bump the `openai-client` lib to 1.10.12, let me know if something is off or doesn't work correctly.

**Comment by shalak** [comment]
@carlrobertoh 

Thanks! The case with 401 is now mapped properly, unfortunately for 404 replies I see:
 
```
{
    "error": {
        "message": "Resource not found",
        "code": "404"
    }
}
```

So it looks more like the old model (note that `code` is a `String`) for non-Azure API.

I think it [would be safer in this place](https://github.com/carlrobertoh/openai-client/commit/ab49d7929f69a725b8512974195b24fe4cfa658d#diff-72c7c283a56d2852a7c179d4c3dbac4ac71f0b0c6243e5a4a6887f03bab369ffR86) if we fallback from any parsing errors and cases where `responseError.getError() == null` into sending `new ErrorDetails(String.format("Unknown API response. Code: %s, Body: %s", response.code(), body.string()))` into the listener (of course body stream must be saved to variable beforhand)

In general we should not trust that API error format will always be the same, it could even by overridden by e.g. local proxy. Looks like various error codes are handled by different services, hence the inconsistency.

**Comment by carlrobertoh** [comment]
It's a bummer that Azure's error responses aren't in sync with OpenAI services, but I suppose that's OK. Also, the property `code` doesn't really make sense to me - is it a response status code or an actual error code? (OpenAI uses it as an actual error code, e.g. `"code": "invalid_api_key"`)

I like the idea of sending the detailed message into the listener, although it makes it bit challenging to extract the JSON body when the need arises. Another option would be putting the actual error body into the `ErrorDetails` and let the consumers handle the formatting.

> In general we should not trust that API error format will always be the same, it could even by overridden by e.g. local proxy. Looks like various error codes are handled by different services, hence the inconsistency.

I agree, altho changing the response payload will be a breaking change for everyone, so I wouldn't worry about that too much. For proxy users, it's up to them to keep the payload structure in sync.

FYI: I released new improvements in version 1.0.13 (might take few minutes until it's available in maven central)

**Comment by shalak** [comment]
> Also, the property code doesn't really make sense to me

🤷 doesn't make sense to me either :) 

> I like the idea of sending the detailed message into the listener, although it makes it bit challenging to extract the JSON body when the need arises. Another option would be putting the actual error body into the ErrorDetails and let the consumers handle the formatting.

The problem is fallback - when all things go in unexpected ways, we see empty chat response from CodeGPT (and sometimes IDE error). That's why I suggested passing the raw JSON to user - in case the API behaves in a way we don't cover in plugin.

**Comment by karolisKTri** [comment]
I see that plugin update came with the Azure OpenAI support. I tried it but couldn't get it working. I get error: "Something went wrong. Please try again later.". How can I debug what's the reason?
Here is how my configuration looks like (but I tried different configurations as well):
![image](https://user-images.githubusercontent.com/130636951/235865622-555f4e2e-f4d4-4070-941e-a970b49405c3.png)
It's not entirely clear what should I put to each field. Maybe there's some example?
Same API works with VSCode plugin.

**Comment by shalak** [comment]
The resource name is not full URL, just the part you censored.

**Comment by carlrobertoh** [comment]
@karolisKTri There's an ongoing [conversation](https://github.com/carlrobertoh/CodeGPT/pull/96) about this.

**Comment by shalak** [comment]
It depends on what kind of authentication you're using (more details can be found in [the PR that introduced Azure Services support](https://github.com/carlrobertoh/CodeGPT/pull/96#issuecomment-1531620303)).

In general - as of today, CodeGPT support only Active Directory authentication (long token with 1988 characters). 

In case you want to contribute the support the other type of auth, in order to use the short token, the openai-client needs a change to set the `api-key` header instead of [the `Authentication: Bearer ...`](https://github.com/carlrobertoh/openai-client/blob/master/src/main/java/ee/carlrobert/openai/client/BaseClient.java#L19). It should of course be toggle-able in CodeGPT itself with a checkbox.

**Comment by carlrobertoh** [comment]
@shalak @karolisKTri I'll make the change today/tomorrow depending on my time.

**Comment by karolisKTri** [comment]
I guess I'm in trouble as I can't use Azure cloud shell to retrieve access token, or is there another simple way to get it?
Anyway I see that new development is planned and I hope that usage of normal API key will be implemented soon.
Thanks.

**Comment by carlrobertoh** [comment]
@karolisKTri Both auth methods have been added in version 1.10.6. Please let me know if something is off or doesn't work properly, couldn't test it myself.

**Comment by renjfk** [comment]
I can confirm both works fine. Although there are two things seemed a bit weird to me unrelated to authentication:

- When GPT stops answering I can still see the icon spinning and stop generating button doesn't disappear also clicking does nothing.
- I'm not sure how this works with OpenAI but Azure OpenAI seems to be unaware of the current context, meaning it doesn't care about chat history.

![image](https://user-images.githubusercontent.com/15015690/236160414-3e2fee93-1d06-4c56-bab7-fe2ffcd6d548.png)

**Comment by karolisKTri** [comment]
I can confirm same behavior.

**Comment by carlrobertoh** [comment]
> When GPT stops answering I can still see the icon spinning and stop generating button doesn't disappear also clicking does nothing.

There might be differences how both services end their stream responses. OpenAI simply says `[DONE]` as a last message.

> I'm not sure how this works with OpenAI but Azure OpenAI seems to be unaware of the current context, meaning it doesn't care about chat history.

That's weird, the logic for building the message context should be the same as for OpenAI services.
I'll try to reproduce it via tests since I don't have access to Azure services.

**Comment by shalak** [comment]
Huh, I don't have any of those issues. When GTP stops answering I see "Regenerate response" button and it does keep track of the conversation.

The way the GPT models keep track of the whole conversation is brutally simple - with every request, [we're sending them the whole conversation](https://github.com/carlrobertoh/CodeGPT/blob/a514983ca3304f74a44080cbaa1efaa234fbf75d/src/main/java/ee/carlrobert/codegpt/client/CompletionRequestProvider.java#L50-L53) it just cannot exceed the token window. It has nothing to do with authentication type... Maybe it's a problem with some settings-mixup, while testing? 

@renjfk @karolisKTri can you check if your `CodeGPTSettings_170.xml` files looks OK? No empty values, no weird characters that may mess with OpenAI's API? (if you were to share them, do not forget to obfuscate the secrets!)

**Comment by renjfk** [comment]
It prolly has to do with model, I'm using `gpt-35-turbo (version 0301)`. What do you use @shalak?

**Comment by shalak** [comment]
I checked with all the Chat Completion models (3.5 & 4.0) - works correctly.

**Comment by renjfk** [comment]
Yes of course otherwise API wouldn't work at all. I'll debug the request/response flow when I have time.

**Comment by renjfk** [comment]
I just had time to debug the plugin so here's how the last message looks like from Azure OpenAI:

```json
{"id":"chatcmpl-7EidTV71v4evAFOWGnkVPzqUQ26TL","object":"chat.completion.chunk","created":1683742527,"model":"gpt-35-turbo","choices":[{"index":0,"finish_reason":"stop","delta":{}}],"usage":null}
```

So `finish_reason` is the indicator for Azure OpenAI as it's also mentioned on the [documentation](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference#chat-completions). I don't know how it works with you @shalak, maybe you are in fact querying directly OpenAI API instead of Azure?

I'll dig a bit more why it doesn't have conversation context.

**Comment by renjfk** [comment]
Actually no need to look further; since it's not able to [receive end signal](https://github.com/carlrobertoh/openai-client/blob/master/src/main/java/ee/carlrobert/openai/client/completion/CompletionEventSourceListener.java#L46) it also doesn't publish completion event thus it doesn't save conversation.

**Comment by renjfk** [comment]
After poking around a bit with a dirty hack it started working perfectly fine:

![image](https://github.com/carlrobertoh/CodeGPT/assets/15015690/30c04479-7063-40c7-b628-bafb8d437a52)

and here's the patch:

```patch
Index: src/main/java/ee/carlrobert/openai/client/completion/CompletionEventSourceListener.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/ee/carlrobert/openai/client/completion/CompletionEventSourceListener.java b/src/main/java/ee/carlrobert/openai/client/completion/CompletionEventSourceListener.java
--- a/src/main/java/ee/carlrobert/openai/client/completion/CompletionEventSourceListener.java	(revision f1294bc444a018b6d01f55b32e58bc50626ce91c)
+++ b/src/main/java/ee/carlrobert/openai/client/completion/CompletionEventSourceListener.java	(date 1683744339781)
@@ -18,7 +18,7 @@
   private static final Logger LOG = LoggerFactory.getLogger(CompletionEventSourceListener.class);
 
   private final CompletionEventListener listeners;
-  private final StringBuilder messageBuilder = new StringBuilder();
+  protected final StringBuilder messageBuilder = new StringBuilder();
 
   public CompletionEventSourceListener(CompletionEventListener listeners) {
     this.listeners = listeners;
Index: src/main/java/ee/carlrobert/openai/client/completion/chat/response/ChatCompletionResponseChoice.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/ee/carlrobert/openai/client/completion/chat/response/ChatCompletionResponseChoice.java b/src/main/java/ee/carlrobert/openai/client/completion/chat/response/ChatCompletionResponseChoice.java
--- a/src/main/java/ee/carlrobert/openai/client/completion/chat/response/ChatCompletionResponseChoice.java	(revision f1294bc444a018b6d01f55b32e58bc50626ce91c)
+++ b/src/main/java/ee/carlrobert/openai/client/completion/chat/response/ChatCompletionResponseChoice.java	(date 1683743872917)
@@ -7,13 +7,21 @@
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class ChatCompletionResponseChoice {
 
+  private final String finishReason;
   private final ChatCompletionResponseChoiceDelta delta;
 
   @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
-  public ChatCompletionResponseChoice(@JsonProperty("delta") ChatCompletionResponseChoiceDelta delta) {
+  public ChatCompletionResponseChoice(
+      @JsonProperty("finish_reason") String finishReason,
+      @JsonProperty("delta") ChatCompletionResponseChoiceDelta delta) {
+    this.finishReason = finishReason;
     this.delta = delta;
   }
 
+  public String getFinishReason() {
+    return finishReason;
+  }
+
   public ChatCompletionResponseChoiceDelta getDelta() {
     return delta;
   }
Index: src/main/java/ee/carlrobert/openai/client/completion/chat/ChatCompletionEventSourceListener.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/ee/carlrobert/openai/client/completion/chat/ChatCompletionEventSourceListener.java b/src/main/java/ee/carlrobert/openai/client/completion/chat/ChatCompletionEventSourceListener.java
--- a/src/main/java/ee/carlrobert/openai/client/completion/chat/ChatCompletionEventSourceListener.java	(revision f1294bc444a018b6d01f55b32e58bc50626ce91c)
+++ b/src/main/java/ee/carlrobert/openai/client/completion/chat/ChatCompletionEventSourceListener.java	(date 1683744339784)
@@ -7,6 +7,7 @@
 import ee.carlrobert.openai.client.completion.CompletionEventSourceListener;
 import ee.carlrobert.openai.client.completion.ErrorDetails;
 import ee.carlrobert.openai.client.completion.chat.response.ChatCompletionResponse;
+import ee.carlrobert.openai.client.completion.chat.response.ChatCompletionResponseChoice;
 
 public class ChatCompletionEventSourceListener extends CompletionEventSourceListener {
 
@@ -14,13 +15,16 @@
     super(listeners);
   }
 
-  protected String getMessage(String data) throws JsonProcessingException {
+  protected static ChatCompletionResponseChoice getChatCompletionResponse(String data)
+      throws JsonProcessingException {
     return new ObjectMapper()
         .readValue(data, ChatCompletionResponse.class)
         .getChoices()
-        .get(0)
-        .getDelta()
-        .getContent();
+        .get(0);
+  }
+
+  protected String getMessage(String data) throws JsonProcessingException {
+    return getChatCompletionResponse(data).getDelta().getContent();
   }
 
   @Override
Index: src/main/java/ee/carlrobert/openai/client/azure/AzureChatCompletionClient.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/ee/carlrobert/openai/client/azure/AzureChatCompletionClient.java b/src/main/java/ee/carlrobert/openai/client/azure/AzureChatCompletionClient.java
--- a/src/main/java/ee/carlrobert/openai/client/azure/AzureChatCompletionClient.java	(revision f1294bc444a018b6d01f55b32e58bc50626ce91c)
+++ b/src/main/java/ee/carlrobert/openai/client/azure/AzureChatCompletionClient.java	(date 1683744355792)
@@ -8,6 +8,8 @@
 import ee.carlrobert.openai.client.completion.CompletionEventSourceListener;
 import ee.carlrobert.openai.client.completion.ErrorDetails;
 import ee.carlrobert.openai.client.completion.chat.ChatCompletionEventSourceListener;
+import okhttp3.sse.EventSource;
+import org.jetbrains.annotations.NotNull;
 
 public class AzureChatCompletionClient extends AzureCompletionClient {
 
@@ -22,6 +24,21 @@
       protected ErrorDetails getErrorDetails(String data) throws JsonProcessingException {
         return new ObjectMapper().readValue(data, AzureApiResponseError.class).getError();
       }
+
+      @Override
+      public void onEvent(@NotNull EventSource eventSource, String id, String type,
+          @NotNull String data) {
+        try {
+          var chatCompletionResponse = getChatCompletionResponse(data);
+          if ("stop".equals(chatCompletionResponse.getFinishReason())) {
+            listeners.onComplete(messageBuilder);
+            return;
+          }
+          super.onEvent(eventSource, id, type, data);
+        } catch (JsonProcessingException e) {
+          throw new RuntimeException("Unable to deserialize payload.", e);
+        }
+      }
     };
   }
 

```

**Comment by shalak** [comment]
I'm 100% sure I'm on Azure OpenAI Services :)

Clearly, it must be a case of the configuration. Mine looks like this:
<img width="692" alt="image" src="https://github.com/carlrobertoh/CodeGPT/assets/1536554/b9b90133-ecf6-4053-b04a-83cdb80ce976">

I also noticed the `"finish_reason":"stop"` on [ChatCompletionEventSourceListener](https://github.com/carlrobertoh/openai-client/blob/master/src/main/java/ee/carlrobert/openai/client/completion/chat/ChatCompletionEventSourceListener.java#L18). But I also do see that [data == `[DONE]`](https://github.com/carlrobertoh/openai-client/blob/master/src/main/java/ee/carlrobert/openai/client/completion/CompletionEventSourceListener.java#L45)

**Comment by renjfk** [comment]
Also @carlrobertoh going back to original topic, I think it still might be interesting to have the option for changing base URL.

There's https://github.com/go-skynet/LocalAI project which let's you simulate OpenAI API so it might be interesting to try it especially when it passes through https://github.com/hwchase17/langchain e.g. creating a chain with web search and let LLM analyze results.

**Comment by carlrobertoh** [comment]
Yeah, I assumed there might be differences how both services end their stream.

@renjfk Would you be so kind to open a PR in `openai-client` repository with your suggested changes? Also, I believe you can override the `onEvent` inside `ChatCompletionEventSourceListener` since OpenAI uses the same logic to end the stream.

@shalak Perhaps the finish behaviour is different for Azure Active Directory vs default auth?

Furthermore, could someone confirm if the text completion works as expected?

**Comment by carlrobertoh** [comment]
> Also @carlrobertoh going back to original topic, I think it still might be interesting to have the option for changing base URL.
> 
> There's https://github.com/go-skynet/LocalAI project which let's you simulate OpenAI API so it might be interesting to try it especially when it passes through https://github.com/hwchase17/langchain e.g. creating a chain with web search and let LLM analyze results.

Yes, that would be really beneficial, I will try to implement as soon as possible (the lib support exist today).

**Comment by shalak** [comment]
> @shalak Perhaps the finish behaviour is different for Azure Active Directory vs default auth?

🤷 I have no means of checking. I might be able to in 2 weeks, once a buddy of mine is back in the office.

> Furthermore, could someone confirm if the text completion works as expected?

I only have 3.5 and 4.0 deployments at the moment, so I can only check chat completion. Again - might be able to in 2 weeks.

**Comment by renjfk** [comment]
> Perhaps the finish behaviour is different for Azure Active Directory vs default auth?

I already did that and it's the same. I think it more has to do with deployment; the only thing I can think of is @shalak uses an earlier deployment and it somehow differs from the one I have.

> Furthermore, could someone confirm if the text completion works as expected?

I can check that today.

**Comment by renjfk** [comment]
Just confirmed text completion works but that doesn't even send end signal via `finish_reason` so I created https://github.com/carlrobertoh/openai-client/pull/1 to rely on okhttp client instead of message content.

**Comment by carlrobertoh** [comment]
@renjfk Added support for custom service/models. Tried several gpt4all models through LocalAI, wasn't the best behaviour but seemed to be working.
